internal/cloud/challenge.go

```diff
@@ -0,0 +1,61 @@
+package cloud
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/containerregistry/runtime/2019-08-15-preview/containerregistry"
+)
+
+type AuthDirective struct {
+	Service string
+	Realm   string
+}
+
+func ChallengeLoginServer(ctx context.Context, loginServerURL string) (*AuthDirective, error) {
+	v2Support := containerregistry.NewV2SupportClient(loginServerURL)
+	challenge, err := v2Support.Check(ctx)
+	// A 401 will also return an error so just check first
+	if !challenge.IsHTTPStatus(401) {
+		if err != nil {
+			return nil, err
+		}
+
+		defer challenge.Body.Close()
+		return nil, fmt.Errorf("registry did not issue a valid AAD challenge, status: %d", challenge.StatusCode)
```

Comment:
maybe something more generic than AAD?

Reply:
Yea, it was lifted from my Azure code since go-containerregistry has similar code but its private.